### PR TITLE
Disable homescreen buttons

### LINKED ISSUE #1602: Allow disabling homescreen buttons
screen shot 2017-09-15 at 10 28 25 am

```diff
diff --git a/source/views/home/edit.js b/source/views/home/edit.js
index 7c10aaf2..2e10733a 100644
--- a/source/views/home/edit.js
+++ b/source/views/home/edit.js
@@ -12,6 +12,7 @@ import {
   StyleSheet,
   Text,
   Platform,
+  Switch,
 } from 'react-native'

 import {saveHomescreenOrder} from '../../flux/parts/homescreen'
@@ -200,6 +201,7 @@ class Row extends React.Component<void, RowProps, RowState> {
         <Text style={[styles.text, {color: this.props.data.tint}]}>
           {this.props.data.title}
         </Text>
+        <Switch />
         {reorderIcon}
       </Animated.View>
     )
```

---

We'd need to store the enabled state in our asyncstorage, and provide it to the Row as appropriate, and listen to the onValueChange callback to set it.

We'd also have to update HomeView to hide buttons as requested.

### PR DESCRIPTION
**Pull request by hannesmcman**
Allow disabling homescreen buttons. Resolves #1602.

### CODE REVIEW

**Comment by hawkrives** [comment]
I'll definitely try this out tonight!

One thing to make sure works (I haven't checked yet): if a new button is added, it needs to be enabled by default, which it probably is I just want to make sure. 

I guess if a button is removed, it should be removed from the disabled list?

**Comment by hawkrives** [comment]
I haven't tried to run this yet, but it looks like you've got a whitelist in here. 

Does this make more sense as a blacklist of disabled buttons than as a whitelist of enabled buttons? I feel like "disabled" will be a less common state than "enabled" (or I hope at least). 

Also, @rye / @drewvolz, do we want to add some analytics for this?

**Comment by drewvolz** [comment]
@hannesmcman I really like your implementation so far 👍

@hawkrives brought up a fun edge case. If you toggle the views that are within [`source/views/views.js`](https://github.com/StoDevX/AAO-React-Native/blob/master/source/views/views.js), it looks like new view objects are not enabled by default.

This is testable by uncommenting the `HelpView` definition

From:
```js
// {
//    type: 'view',
//    view: 'HelpView',
//    title: 'Report A Problem',
//    icon: 'help',
//    tint: c.lavender,
//    gradient: c.purpleToIndigo,
// },
  ```
To:
```js
{
  type: 'view',
  view: 'HelpView',
  title: 'Report A Problem',
  icon: 'help',
  tint: c.lavender,
  gradient: c.purpleToIndigo,
},
  ```

and the `EditHome` list should show up with the `HelpView` row as disabled.

**Comment by drewvolz** [comment]
iOS builds and runs.

<details>
<summary> As for Android, this is what I see:</summary>
<br>

```bash
TypeError: undefined is not an object (evaluating 'this.props.activeViews.includes') 

This error is located at:
    in EditHomeRow (at list.android.js:65)
    in RCTView (at View.js:112)
    in View (at VirtualizedList.js:1369)
    in CellRenderer (at VirtualizedList.js:556)
    in RCTView (at View.js:112)
    in View (at ScrollView.js:748)
    in RCTScrollView (at ScrollView.js:857)
    in ScrollView (at VirtualizedList.js:877)
    in VirtualizedList (at FlatList.js:622)
    in FlatList (at list.android.js:84)
    in EditHomeView (created by Connect(EditHomeView))
    in Connect(EditHomeView) (at SceneView.js:31)
    in SceneView (at CardStack.js:412)
    in RCTView (at View.js:112)
    in View (at CardStack.js:411)
    in RCTView (at View.js:112)
    in View (at CardStack.js:410)
    in RCTView (at View.js:112)
    in View (at createAnimatedComponent.js:134)
    in AnimatedComponent (at Card.js:26)
    in Card (at PointerEventsContainer.js:55)
    in Container (at CardStack.js:454)
    in RCTView (at View.js:112)
    in View (at CardStack.js:383)
    in RCTView (at View.js:112)
    in View (at CardStack.js:382)
    in CardStack (at CardStackTransitioner.js:97)
    in RCTView (at View.js:112)
    in View (at Transitioner.js:192)
    in Transitioner (at CardStackTransitioner.js:49)
    in CardStackTransitioner (at StackNavigator.js:60)
    in Unknown (at createNavigator.js:52)
    in Navigator (at createNavigationContainer.js:210)
    in NavigationContainer (at app.js:101)
    in Provider (at app.js:100)
    in App (at renderApplication.js:35)
    in RCTView (at View.js:112)
    in View (at AppContainer.js:102)
    in RCTView (at View.js:112)
    in View (at AppContainer.js:122)
    in AppContainer (at renderApplication.js:34)
render
    row.android.js:81:43
proxiedMethod
    createPrototypeProxy.js:44:35
finishClassComponent
    ReactNativeFiber-dev.js:1667:86
updateClassComponent
    ReactNativeFiber-dev.js:1659:33
beginWork
    ReactNativeFiber-dev.js:1786:44
performUnitOfWork
    ReactNativeFiber-dev.js:2528:33
workLoop
    ReactNativeFiber-dev.js:2554:141
_invokeGuardedCallback
    ReactNativeFiber-dev.js:73:23
invokeGuardedCallback
    ReactNativeFiber-dev.js:47:40
performWork
    ReactNativeFiber-dev.js:2593:41
batchedUpdates
    ReactNativeFiber-dev.js:2756:116
batchedUpdates
    ReactNativeFiber-dev.js:202:35
batchedUpdatesWithControlledComponents
    ReactNativeFiber-dev.js:209:34
_receiveRootNodeIDEvent
    ReactNativeFiber-dev.js:3559:50
receiveTouches
    ReactNativeFiber-dev.js:3572:64
__callFunction
    MessageQueue.js:299:47
<unknown>
    MessageQueue.js:111:26
__guard
    MessageQueue.js:262:6
callFunctionReturnFlushedQueue
    MessageQueue.js:110:17
```

**Comment by hawkrives** [comment]
I glanced over it again (still haven't run it), and I've got a refactoring I'd apppreciate:

Could you remove the Redux connection from Row?

I'm thinking just connect List to redux (which it already is), then pass two new props to Row: `isEnabled` and `onToggle`; then Row just calls `this.props.onToggle` when the Switch is tapped, and List handles that and does the Redux update, which gives us the new list of disabled views, causing us to update isEnabled. 

Objections, anyone?

**Comment by hannesmcman** [comment]
In https://github.com/StoDevX/AAO-React-Native/pull/2032/commits/5e5ad2b1dc1d5e91ef09595a7d7969f8c1a9c4d9 I changed the whitelist to a blacklist, fixed the issue with new views being disabled by default, removed the redux from Row and made sure that if a view is removed, it is removed from the disabled list.

File: source/flux/parts/homescreen.js [hunk 1]
```diff
@@ -70,71 +70,72 @@ export function saveHomescreenOrder(
...
+  return {type: SAVE_DISABLED_VIEWS, payload: disabledViews}
 }
-export async function loadEnabledViews() {
-  let enabledViews = await storage.getEnabledViews()
+export async function loadDisabledViews() {
+  let disabledViews = await storage.getDisabledViews()
 
-  if (enabledViews.length == 0) {
-    enabledViews = defaultViewOrder
+  if (disabledViews.length == 0) {
```
Comment (hawkrives) [inline source/flux/parts/homescreen.js]: Please use `===` here

File: source/views/home/edit/list.ios.js [hunk 2]
```diff
@@ -61,9 +64,20 @@ class EditHomeView extends React.PureComponent<Props, State> {
...
-    <EditHomeRow active={active} data={data} width={this.state.width} />
-  )
+  renderRow = ({data, active}: {data: ViewType, active: boolean}) => {
+    const enabled = !this.props.inactiveViews.includes(data.view)
+    return (
+      <EditHomeRow
+        active={active}
+        data={data}
+        isEnabled={enabled}
+        onToggle={()=>{
```
Comment (hawkrives) [inline source/views/home/edit/list.ios.js]: We can "simplify" this: here, `onToggle={this.props.onToggleViewDisabled}`, and in Row.js, change `onToggleSwitch` to just call `this.props.onToggle(this.props.data.view)`, like you did before.

**Review by hawkrives — DISMISSED** [review]
Looking good! A few comments; I can elaborate if requested.

**Review by hawkrives — COMMENTED** [review]
👍 Nice!

**Review by hawkrives — APPROVED** [review]
:shipit:

**Comment by rye** [comment]
(In response to https://github.com/StoDevX/AAO-React-Native/pull/2032#issuecomment-355195526)

>do we want to add some analytics for this?

That might be nice.  Every time we update the store, perhaps?  Or just when a button changes state.

**Comment by hawkrives** [comment]
Oh right, I forgot I had noted that. Do you want to open an issue for it?

**Comment by rye** [comment]
Sure...

**Comment by hawkrives** [comment]
I can also open an issue for it

**Comment by hawkrives** [comment]
Then we can have two issues

**Pull request merged**